test/NuGet.Clients.Tests/NuGet.PackageManagement.VisualStudio.Test/CpsPackageReferenceProjectTests.cs

```diff
@@ -4127,6 +4126,107 @@ public async Task GetInstalledAndTransitivePackagesAsync_TransitiveOriginsParamS
+            Assert.Equal(transitivePackagesCount, result.TransitivePackages.Count);
+            Assert.All(result.TransitivePackages, pkg => Assert.Equal(transitiveOriginsCount, pkg.TransitiveOrigins.Count()));
+        }
+
+        [Theory]
+        [InlineData(false, false)]
+        [InlineData(false, true)]
+        [InlineData(true, false)]
+        [InlineData(true, true)]
+        public async Task GetInstalledAndTransitivePackagesAsync_IsInstalledAndTransitiveComputationNeeded_True_Succeed(bool includeTransitivePackages, bool includeTransitiveOrigins)
+        {
+            using var rootDir = new SimpleTestPathContext();
+            await CreatePackagesAsync(rootDir, packageAVersion: "1.0.0");
+            IProjectSystemCache cache = new ProjectSystemCache();
+
+            // Arrange
+            PackageSpec initialProjectSpec = ProjectTestHelpers.GetPackageSpec("MyProject", rootDir.SolutionRoot, framework: "net472", dependencyName: "PackageA");
+            await RestorePackageSpecsAsync(rootDir, output: null, initialProjectSpec);
+
+            CpsPackageReferenceProject project = PrepareCpsRestoredProject(initialProjectSpec, cache);
+
+            // Act
+            ProjectPackages result = await project.GetInstalledAndTransitivePackagesAsync(includeTransitivePackages, includeTransitiveOrigins, CancellationToken.None);
+
+            // Assert
+            var transitivePackagesCount = includeTransitivePackages ? 1 : 0;
+            var transitiveOriginsCount = (includeTransitivePackages && includeTransitiveOrigins) ? 1 : 0;
+
+            Assert.Equal(1, result.InstalledPackages.Count);
+            Assert.Equal(transitivePackagesCount, result.TransitivePackages.Count);
+            Assert.All(result.TransitivePackages, pkg => Assert.Equal(transitiveOriginsCount, pkg.TransitiveOrigins.Count()));
+        }
+
+        [Fact]
+        public async Task GetInstalledAndTransitivePackagesAsync_HasExpectedCache_Succeed()
+        {
+            using var rootDir = new SimpleTestPathContext();
+            await CreatePackagesAsync(rootDir, packageAVersion: "1.0.0");
+            IProjectSystemCache cache = new ProjectSystemCache();
+
+            // Arrange
+            PackageSpec initialProjectSpec = ProjectTestHelpers.GetPackageSpec("MyProject", rootDir.SolutionRoot, framework: "net472", dependencyName: "PackageA");
+            await RestorePackageSpecsAsync(rootDir, output: null, initialProjectSpec);
+
+            CpsPackageReferenceProject project = PrepareCpsRestoredProject(initialProjectSpec, cache);
+
+            // Act
+            ProjectPackages result = await project.GetInstalledAndTransitivePackagesAsync(true, true, CancellationToken.None);
+
+            // Assert
+            Assert.Equal(1, result.InstalledPackages.Count);
+            Assert.Equal(1, result.TransitivePackages.Count);
+            Assert.All(result.TransitivePackages, pkg => Assert.Equal(1, pkg.TransitiveOrigins.Count()));
```

Comment:
I believe these assertions under the same conditions are already covered in other tests above. Please correct me if I'm wrong. If it's true, then we shouldn't be testing them again in another test. Assume they are correct because another test covers this. We can simply set up our test here.

In general, whenever we see Arrange I / Act I / Assert I and Arrange II / Act II / Assert II, ..., etc. it suggests that we're testing more than one thing in a test and that should probably be broken up into multiple tests or we are asserting things we already covered in another test already.

Reply:
Removing these 3 asserts, I was trying to be more clear that the cache was "something" and changes to other state, but we can assume its true since we have tests that cover these asserts.

Reply:
I think those tests are trying to test how the cache evolves over time. e.g. Step 1: Read on Installed tab, Step 2: go to browse tab and so on.

Reply:
@jebriede Yes, this test is trying to test how the cache changes when calling GetInstalledPackages or GIATPA. I didn't specify in the test which user action triggers the GetInstalledPackages or GIATPA, because they can be called from multiple actions (doing a refresh in a tab, changing tabs, installing package) this is just testing that even if the cache was cleared, we can calculate transitive data correctly.

There is a comment here https://github.com/NuGet/NuGet.Client/pull/5117/files#diff-f3327a2a5bc6e2fc4bc816d16cea0eee2f993d91e354fee24676d717efd1a74eR4205, where we are expecting the cache for transitive's to be clear